File: flutter_ui/lib/state/keymap.dart
// The keyboard: turning a real keypress into a chord, and asking the engine
// what that chord means here (docs/07-UI-SPEC.md §15, K-199).
//
// **What lives here and what does not.** This file knows how a Flutter
// `KeyEvent` spells itself — that `LogicalKeyboardKey.pageUp` is the key the
// keymap calls `PageUp`, and that the primary modifier is Cmd on a Mac and Ctrl
// everywhere else. That is platform knowledge, and it belongs on the platform's
// side of the seam. Everything after that is the engine's: which action a chord
// runs, whether the focused panel outranks the app-wide binding, what clashes
// with what, and what the keymap file says. `crates/lumit-bridge/src/api/keymap.rs`
// answers all of those and this file only asks.
//
// The keymap itself is kept by the engine for the session. Its *file* is ours,
// because a keymap is machine-local settings and the workspace file is where
// those live — [KeymapState] stores the engine's JSON verbatim and never looks
// inside it.

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart' show SingleActivator;
import 'package:flutter_rust_bridge/flutter_rust_bridge.dart';

import '../src/rust/api/keymap.dart';
import 'workspace.dart';

/// The keys the keymap names in words, by the logical key Flutter reports.
///
/// Only the ones whose Lumit name differs from what `keyLabel` would give, or
/// where `keyLabel` is unreliable across layouts. Letters, digits and
/// punctuation fall through to the label, which is already what the keymap
/// spells them as.
final Map<LogicalKeyboardKey, String> _namedKeys = {
  LogicalKeyboardKey.space: 'Space',
  LogicalKeyboardKey.pageUp: 'PageUp',
  LogicalKeyboardKey.pageDown: 'PageDown',
  LogicalKeyboardKey.home: 'Home',
  LogicalKeyboardKey.end: 'End',
  LogicalKeyboardKey.enter: 'Enter',
  LogicalKeyboardKey.numpadEnter: 'Enter',
  LogicalKeyboardKey.backspace: 'Backspace',
  LogicalKeyboardKey.delete: 'Delete',
  LogicalKeyboardKey.escape: 'Escape',
  LogicalKeyboardKey.tab: 'Tab',
  LogicalKeyboardKey.arrowUp: 'ArrowUp',
  LogicalKeyboardKey.arrowDown: 'ArrowDown',
  LogicalKeyboardKey.arrowLeft: 'ArrowLeft',
  LogicalKeyboardKey.arrowRight: 'ArrowRight',
  // The marker key is AE's numpad asterisk. On the main rows `*` is Shift+8,
  // which is a different chord and deliberately not folded into this one.
  LogicalKeyboardKey.numpadMultiply: '*',
  LogicalKeyboardKey.numpadAdd: '+',
  LogicalKeyboardKey.numpadSubtract: '-',
  LogicalKeyboardKey.numpadDecimal: '.',
};

/// The modifier keys themselves, which are never a chord on their own — holding
/// Shift is not a shortcut, it is half of one.
final Set<LogicalKeyboardKey> _modifierKeys = {
  LogicalKeyboardKey.shift,
  LogicalKeyboardKey.shiftLeft,
  LogicalKeyboardKey.shiftRight,
  LogicalKeyboardKey.control,
  LogicalKeyboardKey.controlLeft,
  LogicalKeyboardKey.controlRight,
  LogicalKeyboardKey.alt,
  LogicalKeyboardKey.altLeft,
  LogicalKeyboardKey.altRight,
  LogicalKeyboardKey.meta,
  LogicalKeyboardKey.metaLeft,
  LogicalKeyboardKey.metaRight,
  LogicalKeyboardKey.capsLock,
};

/// The keymap's name for one logical key, or null when it has none we can
/// spell — a dead key, or a media key no shortcut should claim.
String? keyName(LogicalKeyboardKey key) {
  if (_modifierKeys.contains(key)) return null;
  final named = _namedKeys[key];
  if (named != null) return named;
  // Function keys are F-and-a-number in both worlds.
  final label = key.keyLabel;
  if (label.isEmpty) return null;
  // A single letter is upper-cased, which is how the keymap stores it; anything
  // else (a digit, punctuation, "F9") is already in its own canonical form.
  return label.length == 1 ? label.toUpperCase() : label;
}

/// The chord [event] spells, in the engine's canonical `Mod+Alt+Shift+Key`
/// form — or null when this keypress cannot be one (a modifier held alone, or
/// a key with no name).
///
/// `Mod` is the platform's primary modifier: Cmd on macOS, Ctrl elsewhere. The
/// keymap and its shareable file stay platform-neutral because of that
/// translation, so a keymap written on a Mac still reads on Windows.
String? chordText(KeyEvent event) {
  final name = keyName(event.logicalKey);
  if (name == null) return null;
  final keys = HardwareKeyboard.instance.logicalKeysPressed;
  bool held(LogicalKeyboardKey a, LogicalKeyboardKey b) =>
      keys.contains(a) || keys.contains(b);
  final primary = defaultTargetPlatform == TargetPlatform.macOS
      ? held(LogicalKeyboardKey.metaLeft, LogicalKeyboardKey.metaRight)
      : held(LogicalKeyboardKey.controlLeft, LogicalKeyboardKey.controlRight);
  final alt = held(LogicalKeyboardKey.altLeft, LogicalKeyboardKey.altRight);
  final shift =
      held(LogicalKeyboardKey.shiftLeft, LogicalKeyboardKey.shiftRight);
  return [
    if (primary) 'Mod',
    if (alt) 'Alt',
    if (shift) 'Shift',
    name,
  ].join('+');
}

/// The logical key each keymap name stands for — [_namedKeys] read backwards.
/// Where two keys share a name (`Enter` is both), the first declared wins, so
/// the main-row key is the one a menu shows.
final Map<String, LogicalKeyboardKey> _keysByName = {
  for (final e in _namedKeys.entries.toList().reversed) e.value: e.key,
};

/// A chord as macOS's own menu bar wants it: a [SingleActivator] for the native
/// `PlatformMenuItem` to draw beside its row (K-244).
///
/// Only the native menu needs this — everywhere else the keyboard is the
/// engine's business and a chord is text. `Mod` becomes Cmd, because macOS is
/// the only place this is asked. Null when the chord names a key we cannot
/// spell as a logical key, so a row shows no shortcut rather than the wrong one.
SingleActivator? activatorForChord(String chord) {
  if (chord.isEmpty) return null;
  // `Mod++` would split into an empty last part; the trailing `+` *is* the key.
  final keyText = chord.endsWith('+') ? '+' : chord.split('+').last;
  final mods = chord.substring(0, chord.length - keyText.length).split('+');
  final key = _keysByName[keyText] ??
      (keyText.length == 1
          ? LogicalKeyboardKey(keyText.toLowerCase().codeUnitAt(0))
          : null);
  if (key == null) return null;
  return SingleActivator(
    key,
    meta: mods.contains('Mod'),
    alt: mods.contains('Alt'),
    shift: mods.contains('Shift'),
  );
}

/// How a chord is *shown* on this machine: `Mod` becomes the symbol or word the
/// platform's own menus use, so a Windows user reads Ctrl and a Mac user reads
/// ⌘. The stored form never changes — only the reading of it.
String chordLabel(String chord) {
  if (chord.isEmpty) return '';
  final mac = defaultTargetPlatform == TargetPlatform.macOS;
  return chord
      .replaceAll('Mod+', mac ? '⌘' : 'Ctrl+')
      .replaceAll('Alt+', mac ? '⌥' : 'Alt+')
      .replaceAll('Shift+', mac ? '⇧' : 'Shift+');
}

/// The live keymap: the table Settings → Keymap draws, and the lookup every
/// keypress goes through.
///
/// Holds no bindings of its own — [groups] and [conflicts] are what the engine
/// last answered, refreshed whenever an edit changes them. Every edit here
/// writes through to the engine first and stores the result second, so the
/// keymap that dispatches and the keymap on disk cannot drift apart.
class KeymapState extends ChangeNotifier {
  KeymapState({Workspace? workspace}) : _workspace = workspace {
    _restore();
    refresh();
  }

  final Workspace? _workspace;

  List<BridgeKeymapGroup> _groups = const [];
  List<BridgeKeyConflict> _conflicts = const [];
  List<BridgeKeyShadow> _shadows = const [];

  /// The whole table, grouped by where each binding is live.
  List<BridgeKeymapGroup> get groups => _groups;

  /// Chords that could fire two actions at once. Empty in the shipped keymap;
  /// the settings page warns when a rebind makes one.
  List<BridgeKeyConflict> get conflicts => _conflicts;

  /// Chords a panel has taken over from an app-wide binding (K-281). Not
  /// clashes — the focused panel wins by a stated rule — but worth saying,
  /// because the app-wide meaning stops working in that one panel. The shipped
  /// keymap carries one on purpose (`L` in the Timeline).
  List<BridgeKeyShadow> get shadows => _shadows;

  /// The search text above the table. Held here rather than in the page so it
  /// survives the page being closed and reopened.
  String _query = '';
  String get query => _query;
  set query(String value) {
    if (_query == value) return;
    _query = value;
    notifyListeners();
  }

  /// The rows matching [query], or the whole table when it is empty. The
  /// filtering is the engine's — it matches the description, the id and the
  /// chord, and only it knows all three.
  List<BridgeKeymapGroup> get visibleGroups {
    if (_query.trim().isEmpty) return _groups;
    final hits = keymapSearch(query: _query.trim())
        .map((b) => '${b.context}/${b.action}')
        .toSet();
    return _groups
        .map((g) => BridgeKeymapGroup(
              context: g.context,
              label: g.label,
              bindings: g.bindings
                  .where((b) => hits.contains('${b.context}/${b.action}'))
                  .toList(),
            ))
        .where((g) => g.bindings.isNotEmpty)
        .toList();
  }

  /// What [event] does while [context] is the focused panel, or null for
  /// nothing bound. This is the dispatch path — one sync call per keypress.
  String? actionFor(BridgeKeyContext context, KeyEvent event) {
    final chord = chordText(event);
    if (chord == null) return null;
    return keymapLookup(context: context, chord: chord);
  }

  /// The chord [action] answers to, as this machine spells it — for a tooltip
  /// that teaches the shortcut (docs/07 §14: every icon control names itself
  /// and its current shortcut). Null when the action has no binding, so a
  /// caller shows the name alone rather than an empty bracket.
  ///
  /// Read from the table the engine last gave us rather than by asking per
  /// hover: this is drawn on every tooltip and the answer only changes when a
  /// rebind lands, which is exactly when [_adopt] refreshes it.
  String? chordFor(String action) {
    final raw = rawChordFor(action);
    return raw == null ? null : chordLabel(raw);
  }

  /// The same binding in the engine's own spelling (`Mod+Shift+Z`), for the one
  /// caller that needs the parts rather than the reading: the macOS native menu,
  /// which wants a [SingleActivator].
  String? rawChordFor(String action) {
    for (final group in _groups) {
      for (final binding in group.bindings) {
        if (binding.action == action && binding.chord.isNotEmpty) {
          return binding.chord;
        }
      }
    }
    return null;
  }

  /// Re-read the table, the conflicts and the shadows from the engine.
  void refresh() {
    _groups = keymapGroups();
    _conflicts = keymapConflicts();
    _shadows = keymapShadows();
    notifyListeners();
  }

  void _adopt(List<BridgeKeymapGroup> groups) {
    _groups = groups;
    _conflicts = keymapConflicts();
    _shadows = keymapShadows();
    _store();
    notifyListeners();
  }

  /// Point an action at a new chord. Returns null on success, or the engine's
  /// own words when the text was not a chord — which the page shows rather
  /// than swallowing.
  Future<String?> rebind(
      BridgeKeyContext context, String action, String chord) async {
    try {
      _adopt(await keymapRebind(
          context: context, action: action, chord: chord));
      return null;
    } on AnyhowException catch (e) {
      return e.message;
    }
  }

  /// Leave an action with no chord at all.
  Future<void> unbind(BridgeKeyContext context, String action) async {
    _adopt(await keymapUnbind(context: context, action: action));
  }

  /// Put one row back to what the shipped keymap gives it.
  Future<void> resetBinding(BridgeKeyContext context, String action) async {
    _adopt(await keymapResetBinding(context: context, action: action));
  }

  /// Replace the whole keymap with a shipped preset.
  Future<void> loadPreset(BridgeKeymapPreset preset) async {
    _adopt(await keymapLoadPreset(preset: preset));
  }

  /// The whole keymap as text, for "Export keymap…". The same format the
  /// workspace stores, so a file a user shares is a file that restores.
  String toJson() => keymapToJson();

  /// Take a keymap from text — an imported file, or the stored blob. Returns
  /// null on success or the engine's words when the text is not a keymap.
  Future<String?> fromJson(String json) async {
    try {
      _adopt(await keymapFromJson(json: json));
      return null;
    } on AnyhowException catch (e) {
      return e.message;
    }
  }

  /// Hand the engine whatever the workspace stored, if anything. A blob that
  /// no longer parses is ignored rather than fatal: the session starts on the
  /// shipped defaults, which is a working keyboard, and the next edit rewrites
  /// the store.
  void _restore() {
    final stored = _workspace?.keymapJson;
    if (stored == null || stored.isEmpty) return;
    try {
      keymapFromJson(json: stored);
    } catch (_) {
      // Defaults it is.
    }
  }

  void _store() {
    final workspace = _workspace;
    if (workspace == null) return;
    workspace.setKeymapJson(keymapToJson());
  }
}

File: crates/lumit-bridge/src/api/keymap.rs
//! The remappable keymap: the table Settings → Keymap draws, and the lookup
//! every keypress goes through (docs/07-UI-SPEC.md §15, K-199).
//!
//! # In plain terms
//!
//! A *chord* is a key with its modifiers — `Space`, `Mod+D`, `Shift+F3`. A
//! *context* is where you are: the whole app, or one focused panel. A *binding*
//! ties a chord in a context to an *action*, named by a stable string like
//! `"playback.toggle"`. Pressing a key asks [`keymap_lookup`] which action that
//! chord means where you are, and the frontend runs it.
//!
//! **Why the model lives here and not in Dart.** Everything that has to be
//! *decided* — what a chord means, whether the focused panel outranks the
//! app-wide binding, whether two bindings clash, what the shareable file says —
//! is a rule, and rules live in the engine (K-181). The frontend turns a real
//! key event into chord text, draws the table, and forwards the edits. It holds
//! no opinion about any of it. `lumit-keymap` is the crate that knows; this
//! module is its window.
//!
//! **Where it is kept.** In memory here for the session, behind one lock. The
//! *file* is the frontend's, because a keymap is machine-local settings and the
//! frontend already owns that file (the workspace JSON): [`keymap_to_json`]
//! hands out the whole map as text and [`keymap_from_json`] takes it back, and
//! Dart never looks inside the string it stored. Same split as the shareable
//! export a user mails to a friend — one format, two reasons to write it.

use flutter_rust_bridge::frb;
use lumit_keymap::{ActionId, Chord, KeyContext, Keymap};
use std::sync::{Mutex, OnceLock};

use crate::api::BridgeError;

/// Where a binding is live. Mirrors `lumit_keymap::KeyContext` across the seam.
#[frb(non_opaque)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BridgeKeyContext {
    Global,
    Tools,
    Project,
    Timeline,
    Viewer,
    Graph,
    Panels,
    Effects,
}

impl From<BridgeKeyContext> for KeyContext {
    fn from(c: BridgeKeyContext) -> Self {
        match c {
            BridgeKeyContext::Global => KeyContext::Global,
            BridgeKeyContext::Tools => KeyContext::Tools,
            BridgeKeyContext::Project => KeyContext::Project,
            BridgeKeyContext::Timeline => KeyContext::Timeline,
            BridgeKeyContext::Viewer => KeyContext::Viewer,
            BridgeKeyContext::Graph => KeyContext::Graph,
            BridgeKeyContext::Panels => KeyContext::Panels,
            BridgeKeyContext::Effects => KeyContext::Effects,
        }
    }
}

impl From<KeyContext> for BridgeKeyContext {
    fn from(c: KeyContext) -> Self {
        match c {
            KeyContext::Global => BridgeKeyContext::Global,
            KeyContext::Tools => BridgeKeyContext::Tools,
            KeyContext::Project => BridgeKeyContext::Project,
            KeyContext::Timeline => BridgeKeyContext::Timeline,
            KeyContext::Viewer => BridgeKeyContext::Viewer,
            KeyContext::Graph => BridgeKeyContext::Graph,
            KeyContext::Panels => BridgeKeyContext::Panels,
            KeyContext::Effects => BridgeKeyContext::Effects,
        }
    }
}

/// One row of the Settings → Keymap table: what the action is called, what it
/// is called internally, and the chord that runs it.
#[frb(non_opaque)]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BridgeKeyBinding {
    pub context: BridgeKeyContext,
    /// The stable id, e.g. `"playback.toggle"` — what the frontend switches on.
    pub action: String,
    /// What the table shows in its left-hand column, e.g. "Play or pause".
    pub description: String,
    /// The chord in its canonical text form, e.g. `"Mod+Shift+P"`. Empty when
    /// the action is currently unbound, which is a state the table shows
    /// rather than hides. One chord per row — K-200 settled that no shipped
    /// action carries two, so a list here would be structure with nothing to
    /// hold.
    pub chord: String,
}

/// One context's worth of rows, with the heading to put above them — the shape
/// the settings page draws top to bottom.
#[frb(non_opaque)]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BridgeKeymapGroup {
    pub context: BridgeKeyContext,
    /// The heading, e.g. "Timeline" or "Anywhere".
    pub label: String,
    pub bindings: Vec<BridgeKeyBinding>,
}

/// One chord that could fire more than one action where both are live — what
/// the page warns about, and what the user resolves by rebinding one of them.
#[frb(non_opaque)]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BridgeKeyConflict {
    pub chord: String,
    /// The competing actions, already described for display.
    pub actions: Vec<String>,
}

/// One chord a panel takes over from an app-wide binding (K-281).
///
/// **Not a clash.** The focused panel gets first refusal and the app-wide
/// binding is the fallback, so the chord runs exactly one action and which one
/// is never in doubt. It is reported because the app-wide meaning does stop
/// working in that one panel, and somebody reading their keymap should be able
/// to see that rather than discover it by pressing the key.
#[frb(non_opaque)]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BridgeKeyShadow {
    pub chord: String,
    /// Where the takeover applies, e.g. "Timeline".
    pub context: String,
    /// What the chord does there, described for display.
    pub action: String,
    /// What it does everywhere else.
    pub shadowed: String,
}

/// Which shipped keymap to load wholesale.
#[frb(non_opaque)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BridgeKeymapPreset {
    /// Lumit's own defaults (docs/07 §15).
    Lumit,
    /// The After Effects muscle-memory alternate.
    AfterEffects,
}

/// The session's keymap. One window means one keymap, and every call here is a
/// handful of string comparisons, so a plain mutex is the whole story — it is
/// never held across a render, an await or the document lock.
static KEYMAP: OnceLock<Mutex<Keymap>> = OnceLock::new();

/// Run `f` against the session keymap, starting from the shipped default the
/// first time. A poisoned lock is recovered rather than propagated: a keymap is
/// settings, and losing every shortcut because some other call panicked would
/// be a far worse outcome than carrying on with the map as it stands.
#[frb(ignore)]
fn with_keymap<R>(f: impl FnOnce(&mut Keymap) -> R) -> R {
    let mutex = KEYMAP.get_or_init(|| Mutex::new(lumit_keymap::default_keymap()));
    let mut guard = mutex.lock().unwrap_or_else(|poison| poison.into_inner());
    f(&mut guard)
}

#[frb(ignore)]
fn row(km: &Keymap, context: KeyContext, action: &ActionId) -> BridgeKeyBinding {
    BridgeKeyBinding {
        context: context.into(),
        action: action.0.clone(),
        description: action.description(),
        chord: km
            .binding_for(context, action)
            .map(ToString::to_string)
            .unwrap_or_default(),
    }
}

/// Every binding, grouped by context in the order the page lists them — the
/// whole table in one call.
///
/// Rebuilt from the map each time rather than cached: it is a few hundred short
/// strings, it is read once when the page opens and once per edit, and a cache
/// here would be one more thing that can disagree with the truth.
#[frb(sync)]
#[must_use]
pub fn keymap_groups() -> Vec<BridgeKeymapGroup> {
    with_keymap(|km| {
        KeyContext::ALL
            .iter()
            .map(|context| {
                // One row per action in this context, in the map's own order,
                // which is the order docs/07 §15 lists them.
                let mut actions: Vec<ActionId> = Vec::new();
                for b in &km.bindings {
                    if b.context == *context && !actions.contains(&b.action) {
                        actions.push(b.action.clone());
                    }
                }
                BridgeKeymapGroup {
                    context: (*context).into(),
                    label: context.label().to_string(),
                    bindings: actions.iter().map(|a| row(km, *context, a)).collect(),
                }
            })
            .filter(|g| !g.bindings.is_empty())
            .collect()
    })
}

/// The rows matching `query` across every context — the search box above the
/// table. An empty query matches everything.
#[frb(sync)]
#[must_use]
pub fn keymap_search(query: String) -> Vec<BridgeKeyBinding> {
    with_keymap(|km| {
        km.search(&query)
            .into_iter()
            .map(|b| BridgeKeyBinding {
                context: b.context.into(),
                action: b.action.0.clone(),
                description: b.action.description(),
                chord: b.chord.to_string(),
            })
            .collect()
    })
}

/// Every chord that could fire two actions at once, described for display.
/// Empty when the keymap is unambiguous, which is the shipped state.
#[frb(sync)]
#[must_use]
pub fn keymap_conflicts() -> Vec<BridgeKeyConflict> {
    with_keymap(|km| {
        km.conflicts()
            .into_iter()
            .map(|c| BridgeKeyConflict {
                chord: c.chord.to_string(),
                actions: c.actions.iter().map(ActionId::description).collect(),
            })
            .collect()
    })
}

/// Every chord a panel takes over from an app-wide binding, described for
/// display (K-281). Said out loud beside the table rather than flagged as
/// something to fix — the shipped default carries one on purpose (`L`).
#[frb(sync)]
#[must_use]
pub fn keymap_shadows() -> Vec<BridgeKeyShadow> {
    with_keymap(|km| {
        km.shadows()
            .into_iter()
            .map(|s| BridgeKeyShadow {
                chord: s.chord.to_string(),
                context: s.context.label().to_string(),
                action: s.action.description(),
                shadowed: s.shadowed.description(),
            })
            .collect()
    })
}

/// What `chord` does while `context` is focused, or `None` for nothing bound.
///
/// This is the dispatch path: every keypress the frontend sees becomes chord
/// text and comes through here. Sync and lock-light for that reason — it is a
/// linear scan of a few hundred bindings, which is nothing beside the frame it
/// happens in, and it must never be the reason a keypress feels late.
#[frb(sync)]
#[must_use]
pub fn keymap_lookup(context: BridgeKeyContext, chord: String) -> Option<String> {
    let Ok(parsed) = chord.parse::<Chord>() else {
        return None;
    };
    with_keymap(|km| km.lookup(context.into(), &parsed).map(|a| a.0.clone()))
}

/// Point an action at a new chord, and hand back the table as it now stands.
///
/// Rejects only chord text that is not a chord; a chord another action already
/// holds is accepted deliberately, because refusing it would make swapping two
/// actions' keys impossible. Within one context the previous owner is left
/// unbound and its row goes blank; a panel-scoped binding taking an app-wide
/// chord leaves both alive, the panel's winning where it is focused, and
/// [`keymap_shadows`] says so (K-281).
pub fn keymap_rebind(
    context: BridgeKeyContext,
    action: String,
    chord: String,
) -> Result<Vec<BridgeKeymapGroup>, BridgeError> {
    let parsed = chord
        .parse::<Chord>()
        .map_err(|e| BridgeError::InvalidKeyChord(e.to_string()))?;
    with_keymap(|km| km.rebind_action(context.into(), &ActionId(action), parsed));
    Ok(keymap_groups())
}

/// Leave an action with no chord at all, and hand back the table.
pub fn keymap_unbind(context: BridgeKeyContext, action: String) -> Vec<BridgeKeymapGroup> {
    let action = ActionId(action);
    let context: KeyContext = context.into();
    with_keymap(|km| km.unbind_action(context, &action));
    keymap_groups()
}

/// Put one action back to the chord the shipped default gives it, and hand
/// back the table. Nothing else in the map is touched — this is the per-row
/// reset, not [`keymap_load_preset`]. Always the Lumit default: "reset" on a
/// settings row means "what the app ships with", whichever preset was loaded
/// since.
pub fn keymap_reset_binding(context: BridgeKeyContext, action: String) -> Vec<BridgeKeymapGroup> {
    let action = ActionId(action);
    let ctx: KeyContext = context.into();
    match lumit_keymap::default_keymap()
        .binding_for(ctx, &action)
        .cloned()
    {
        Some(chord) => with_keymap(|km| km.rebind_action(ctx, &action, chord)),
        // The default does not bind it, so "reset" means unbound.
        None => return keymap_unbind(context, action.0),
    }
    keymap_groups()
}

/// Replace the whole keymap with a shipped preset, and hand back the table.
pub fn keymap_load_preset(preset: BridgeKeymapPreset) -> Vec<BridgeKeymapGroup> {
    let shipped = preset_map(preset);
    with_keymap(|km| *km = shipped);
    keymap_groups()
}

#[frb(ignore)]
fn preset_map(preset: BridgeKeymapPreset) -> Keymap {
    match preset {
        BridgeKeymapPreset::Lumit => lumit_keymap::default_keymap(),
        BridgeKeymapPreset::AfterEffects => lumit_keymap::after_effects_preset(),
    }
}

/// The whole keymap as JSON — what the frontend stores between sessions and
/// what "Export keymap…" writes to a file the user can share. One format for
/// both, so a keymap that survives a restart is the same keymap that travels.
#[frb(sync)]
#[must_use]
pub fn keymap_to_json() -> String {
    with_keymap(|km| serde_json::to_string_pretty(km).unwrap_or_else(|_| String::from("{}")))
}

/// Take a keymap back from JSON — a restored session, or an imported file — and
/// hand back the table.
///
/// Rejects anything that is not a keymap rather than half-applying it, so a
/// corrupt stored blob or somebody else's JSON leaves the current map alone.
///
/// **Laid over the shipped defaults, not swapped for them** (K-302). A file
/// only knows the actions that existed when it was written, and it used to
/// replace the map whole — so every action added since was left with no chord
/// at all for anyone who had ever saved a keymap. That is how `Ctrl+C` came to
/// do nothing in a build whose every test passed. An action the file names
/// keeps the file's chord and an action it deliberately unbound stays unbound;
/// only the ones it never heard of take their default.
pub fn keymap_from_json(json: String) -> Result<Vec<BridgeKeymapGroup>, BridgeError> {
    let parsed: Keymap =
        serde_json::from_str(&json).map_err(|e| BridgeError::InvalidKeymapFile(e.to_string()))?;
    if parsed.bindings.is_empty() {
        return Err(BridgeError::InvalidKeymapFile(
            "the file holds no bindings".to_string(),
        ));
    }
    with_keymap(|km| *km = lumit_keymap::with_new_defaults(parsed));
    Ok(keymap_groups())
}
